File: bindings/zig/src/uhppote/ut0311.zig
const std = @import("std");
const errors = @import("errors.zig");
const network = @import("network.zig");

const READ_TIMEOUT = 2500 * std.time.us_per_ms;
const WRITE_TIMEOUT = 1000 * std.time.us_per_ms;
const BUFFER_SIZE = 1024;

pub const Controller = struct {
    controller: u32,
    address: [:0]const u8,
    transport: [:0]const u8,
};

var bindAddr = network.EndPoint{
    .address = network.Address{ .ipv4 = network.Address.IPv4.any },
    .port = 0,
};

var broadcastAddr = network.EndPoint{
    .address = network.Address{ .ipv4 = network.Address.IPv4.broadcast },
    .port = 60000,
};

var listenAddr = network.EndPoint{
    .address = network.Address{ .ipv4 = network.Address.IPv4.any },
    .port = 60001,
};

var debug: bool = false;

pub fn set_bind_address(addr: [:0]const u8) !void {
    var address: network.Address = network.Address{ .ipv4 = network.Address.IPv4.any };
    var port: u16 = 0;
    var it = std.mem.split(u8, addr, ":");

    if (it.next()) |v| {
        address = network.Address{ .ipv4 = try network.Address.IPv4.parse(v) };
    }

    if (it.next()) |v| {
        port = try std.fmt.parseUnsigned(u16, v, 10);
    }

    bindAddr = network.EndPoint{
        .address = address,
        .port = port,
    };
}

pub fn set_broadcast_address(addr: [:0]const u8) !void {
    var address: network.Address = network.Address{ .ipv4 = network.Address.IPv4.broadcast };
    var port: u16 = 60000;
    var it = std.mem.split(u8, addr, ":");

    if (it.next()) |v| {
        address = network.Address{ .ipv4 = try network.Address.IPv4.parse(v) };
    }

    if (it.next()) |v| {
        port = try std.fmt.parseUnsigned(u16, v, 10);
    }

    broadcastAddr = network.EndPoint{
        .address = address,
        .port = port,
    };
}

pub fn set_listen_address(addr: [:0]const u8) !void {
    var address: network.Address = network.Address{ .ipv4 = network.Address.IPv4.any };
    var port: u16 = 60001;
    var it = std.mem.split(u8, addr, ":");

    if (it.next()) |v| {
        address = network.Address{ .ipv4 = try network.Address.IPv4.parse(v) };
    }

    if (it.next()) |v| {
        port = try std.fmt.parseUnsigned(u16, v, 10);
    }

    listenAddr = network.EndPoint{
        .address = address,
        .port = port,
    };
}

pub fn set_debug(v: bool) !void {
    debug = v;
}

pub fn broadcast(packet: [64]u8, allocator: std.mem.Allocator) ![][64]u8 {
    var socket = try network.Socket.create(.ipv4, .udp);
    
    defer socket.close();

    try socket.setBroadcast(true);
    try socket.setWriteTimeout(WRITE_TIMEOUT);
    try socket.bind(bindAddr);

    const N = try socket.sendTo(broadcastAddr, &packet);
    if (debug) {
        std.debug.print("   ... sent {any} bytes\n", .{N});
    }

    dump(packet);

    return try read_all(&socket, allocator);
}

pub fn send(controller: Controller, packet: [64]u8, allocator: std.mem.Allocator) ![64]u8 {
    if (!std.mem.eql(u8, controller.address, "") and std.mem.eql(u8, controller.transport, "tcp")) {
        return tcp_sendto(packet, controller.address, allocator);
    }

    if (!std.mem.eql(u8, controller.address, "")) {
        return udp_sendto(packet, controller.address, allocator);
    }

    return udp_broadcast_to(packet, allocator);
}

fn udp_broadcast_to(packet: [64]u8, allocator: std.mem.Allocator) ![64]u8 {
    var socket = try network.Socket.create(.ipv4, .udp);
    
    defer socket.close();

    try socket.setBroadcast(true);
    try socket.setWriteTimeout(WRITE_TIMEOUT);
    try socket.bind(bindAddr);

    const N = try socket.sendTo(broadcastAddr, &packet);
    if (debug) {
        std.debug.print("   ... sent {any} bytes\n", .{N});
    }

    dump(packet);

    // set-ip does not return a response
    if (packet[1] == 0x96) { 
        return [_]u8{0} ** 64;
    } 

    return try read(&socket, allocator);    
}

fn udp_sendto(packet: [64]u8, address:[:0]const u8, allocator: std.mem.Allocator) ![64]u8 {
    const addr = try resolve(address);
    var socket = try network.Socket.create(.ipv4, .udp);
    
    defer socket.close();

    try socket.setWriteTimeout(WRITE_TIMEOUT);
    try socket.setReadTimeout(READ_TIMEOUT);
    try socket.bind(bindAddr);
    try socket.connect(addr);

    const N = try socket.send(&packet);
    if (debug) {
        std.debug.print("   ... sent {any} bytes\n", .{N});
    }

    dump(packet);

    // set-ip does not return a response
    if (packet[1] == 0x96) { 
        return [_]u8{0} ** 64;
    } 

    return try read(&socket, allocator);    
}

fn tcp_sendto(packet: [64]u8, address:[:0]const u8, _: std.mem.Allocator) ![64]u8 {
    const addr = try resolve(address);
    var socket = try network.Socket.create(.ipv4, .tcp);
    
    defer socket.close();

    try socket.setWriteTimeout(WRITE_TIMEOUT);
    try socket.setReadTimeout(READ_TIMEOUT);
    try socket.bind(bindAddr);
    try socket.connect(addr);

    const N = try socket.send(&packet);
    if (debug) {
        std.debug.print("   ... sent {any} bytes\n", .{N});
    }

    dump(packet);

    // set-ip does not return a response
    if (packet[1] == 0x96) { 
        return [_]u8{0} ** 64;
    } 


    var msg: [BUFFER_SIZE]u8 = undefined;

    if (socket.receive(&msg)) |NN| {
        if (debug) {
            std.debug.print("   ... received {any} bytes\n", .{NN});
        }

        if (NN != 64) {
            return errors.UhppotedError.InvalidReply;
        }

        dump(msg[0..64].*);

        return msg[0..64].*;
    } else |err| switch (err) {
        error.WouldBlock => {
            return errors.UhppotedError.NoReply;
        },
        else => {
            return err;
        },
    }
}

pub fn listen(queue: *std.fifo.LinearFifo([64]u8,.Dynamic), _: std.mem.Allocator) !void {
    var socket = try network.Socket.create(.ipv4, .udp);
    defer socket.close();

    try socket.bind(listenAddr);

    while (true) {
        var buffer: [BUFFER_SIZE]u8 = undefined;
        if (socket.receiveFrom(&buffer)) |reply| {
            if (debug) {
                std.debug.print("   ... received {any} bytes\n", .{reply.numberOfBytes});
            }

            if (reply.numberOfBytes == 64) {
                dump(buffer[0..64].*);

                if (queue.writeItem(buffer[0..64].*)) {
                } else |err| {
                    std.debug.print("\n   *** WARN   {any}\n", .{err});                
                }
            }
        } else |err| {
            std.debug.print("{any}\n", .{err});
            break;
        }
    }
}

fn read_all(socket: *network.Socket, allocator: std.mem.Allocator) ![][64]u8 {
    try socket.setReadTimeout(100 * std.time.us_per_ms);

    const start = std.time.milliTimestamp() * std.time.us_per_ms;
    var replies = std.ArrayList([64]u8).init(allocator);
    defer replies.deinit();

    while (true) {
        var msg: [BUFFER_SIZE]u8 = undefined;
        if (socket.receiveFrom(&msg)) |reply| {
            if (debug) {
                std.debug.print("   ... received {any} bytes\n", .{reply.numberOfBytes});
            }

            if (reply.numberOfBytes == 64) {
                const slice: [64]u8 = msg[0..64].*;
                if (replies.append(slice)) {
                    dump(msg[0..64].*);
                } else |err| {
                    std.debug.print("{any}\n", .{err});
                }
            }
        } else |err| switch (err) {
            error.WouldBlock => {
                const dt = std.time.milliTimestamp()*std.time.us_per_ms - start;

                if (dt >= READ_TIMEOUT) {
                    break;
                }
            },
            else => {
                std.debug.print("{any}\n", .{err});
                break;
            },
        }
    }

    return replies.toOwnedSlice();
}

fn read(socket: *network.Socket, _: std.mem.Allocator) ![64]u8 {
    try socket.setReadTimeout(READ_TIMEOUT);

    while (true) {
        var msg: [BUFFER_SIZE]u8 = undefined;

        if (socket.receiveFrom(&msg)) |reply| {
            if (debug) {
                std.debug.print("   ... received {any} bytes\n", .{reply.numberOfBytes});
            }

            if (reply.numberOfBytes == 64) {
                dump(msg[0..64].*);

                return msg[0..64].*;
            }
        } else |err| switch (err) {
            error.WouldBlock => {
                return errors.UhppotedError.NoReply;
            },
            else => {
                return err;
            },
        }
    }
}

fn resolve(addr: [:0]const u8) !network.EndPoint {
    var address: network.Address = network.Address{ .ipv4 = network.Address.IPv4.any };
    var port: u16 = 60000;
    var it = std.mem.split(u8, addr, ":");

    if (it.next()) |v| {
        address = network.Address{ .ipv4 = try network.Address.IPv4.parse(v) };
    }

    if (it.next()) |v| {
        port = try std.fmt.parseUnsigned(u16, v, 10);
    }

    return network.EndPoint{
        .address = address,
        .port = port,
    };
}

fn dump(packet: [64]u8) void {
    if (debug) {
        const offsets = [_]usize{ 0, 16, 32, 48 };
        for (offsets) |ix| {
            const f = "   {x:0>2} {x:0>2} {x:0>2} {x:0>2} {x:0>2} {x:0>2} {x:0>2} {x:0>2}";
            const u = packet[ix .. ix + 8];
            const v = packet[ix + 8 .. ix + 16];
            std.debug.print(f, .{ u[0], u[1], u[2], u[3], u[4], u[5], u[6], u[7] });
            std.debug.print("  ", .{});
            std.debug.print(f, .{ v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7] });
            std.debug.print("\n", .{});
        }

        std.debug.print("\n", .{});
    }
}

File: bindings/go/uhppote/UT0311.go
package uhppote

import (
    "fmt"
    "net"
    "net/netip"
    "time"
    "syscall"
)

const CONNECT_TIMEOUT = 1000 * time.Millisecond
const READ_TIMEOUT = 5000 * time.Millisecond
const WRITE_TIMEOUT = 1000 * time.Millisecond

var bindAddr netip.AddrPort = netip.MustParseAddrPort("0.0.0.0:0")
var broadcastAddr *net.UDPAddr = resolveAddr("255.255.255.255:60000")
var listenAddr *net.UDPAddr = resolveAddr("0.0.0.0:60001")
var debug bool = false

func SetBindAddr(addr string) {
    address := netip.MustParseAddr(addr)
    bindAddr = netip.AddrPortFrom(address, 0)
}

func SetBroadcastAddr(addr string) {
    broadcastAddr = resolveAddr(addr)
}

func SetListenAddr(addr string) {
    listenAddr = resolveAddr(addr)
}

func SetDebug(enabled bool) {
    debug = enabled
}

func broadcast(request []byte) ([][]byte, error) {
    dump(request)

    bind := net.UDPAddrFromAddrPort(bindAddr)
    if bind == nil {
        bind = &net.UDPAddr{
            IP:   net.IPv4(0, 0, 0, 0),
            Port: 0,
            Zone: "",
        }
    }

    socket, err := net.ListenUDP("udp", bind)
    if err != nil {
        return nil, err
    }

    defer socket.Close()

    if err := socket.SetWriteDeadline(time.Now().Add(WRITE_TIMEOUT)); err != nil {
        return nil, err
    }

    if err := socket.SetReadDeadline(time.Now().Add(READ_TIMEOUT)); err != nil {
        return nil, fmt.Errorf("Failed to set UDP read timeout [%v]", err)
    }

    if _, err := socket.WriteToUDP(request, broadcastAddr); err != nil {
        return nil, err
    }

    return readAll(socket)
}

func send(controller Controller, request []byte) ([]byte, error) {
    if controller.Address != "" && controller.Transport == "tcp" {
        if addr, err := netip.ParseAddrPort(controller.Address); err != nil {
            return nil, err
        } else {
            return tcpSendTo(request, net.TCPAddrFromAddrPort(addr))
        }
    } 

    if controller.Address != "" && controller.Transport == "udp" {
        if addr, err := netip.ParseAddrPort(controller.Address); err != nil {
            return nil, err
        } else {
            return udpSendTo(request, net.UDPAddrFromAddrPort(addr))
        }
    } 
      
    return udpBroadcastTo(request)
}

func udpBroadcastTo(request []byte) ([]byte, error) {
    dump(request)

    bind := net.UDPAddrFromAddrPort(bindAddr)
    if bind == nil {
        bind = &net.UDPAddr{
            IP:   net.IPv4(0, 0, 0, 0),
            Port: 0,
            Zone: "",
        }
    }

    if connection, err := net.ListenUDP("udp", bind); err != nil {
        return nil, fmt.Errorf("error creating UDP socket (%v)", err)
    } else if connection == nil {
        return nil, fmt.Errorf("open() created invalid UDP socket (%v)", connection)
    } else {
        defer connection.Close()

        deadline := time.Now().Add(READ_TIMEOUT)
        if err := connection.SetDeadline(deadline); err != nil {
            return nil, err
        }

        if _, err := connection.WriteToUDP(request, broadcastAddr); err != nil {
            return nil, fmt.Errorf("failed to write to UDP socket [%v]", err)
        }

        // NTS: set-ip doesn't return a reply so fake it
        if request[1] == 0x96 {
            reply := make([]byte, 64)

            copy(reply, request[:8])
            reply[8] = 0x01

            return reply, nil
        }

        for {
            reply := make([]byte, 2048)

            if N, _, err := connection.ReadFromUDP(reply); err != nil {
                return nil, err
            } else {
                dump(reply[:N])

                return reply[:N], err
            }
        }
    }
}

func udpSendTo(request []byte, addr *net.UDPAddr) ([]byte, error) {
    dump(request)

    address := fmt.Sprintf("%v", addr)
    bind := net.UDPAddrFromAddrPort(bindAddr)

    if bind == nil {
        bind = &net.UDPAddr{
            IP:   net.IPv4(0, 0, 0, 0),
            Port: 0,
            Zone: "",
        }
    }

    dialer := net.Dialer{
        Deadline:  time.Now().Add(CONNECT_TIMEOUT),
        LocalAddr: bind,
        Control: func(network, address string, connection syscall.RawConn) (err error) {
            var operr error

            f := func(fd uintptr) {
                operr = setSocketOptions(fd)
            }

            if err := connection.Control(f); err != nil {
                return err
            } else {
                return operr
            }
        },
    }

    if connection, err := dialer.Dial("udp4", address); err != nil {
        return nil, err
    } else if connection == nil {
        return nil, fmt.Errorf("invalid UDP socket (%v)", connection)
    } else {
        defer connection.Close()

        deadline := time.Now().Add(READ_TIMEOUT)
        buffer := make([]byte, 1024)

        if err := connection.SetDeadline(deadline); err != nil {
            return nil, err
        }

        if _, err := connection.Write(request); err != nil {
            return nil, fmt.Errorf("failed to write to UDP socket [%v]", err)
        }

        // NTS: set-ip doesn't return a reply
        if request[1] == 0x96 {
            return nil, nil
        }

        if N, err := connection.Read(buffer); err != nil {
            return nil, err
        } else {
            dump(buffer[:N])

            return buffer[0:N], nil
        }
    }
}

func tcpSendTo(request []byte, addr *net.TCPAddr) ([]byte, error) {
    dump(request)

    address := fmt.Sprintf("%v", addr)
    bind := net.TCPAddrFromAddrPort(bindAddr)

    if bind == nil {
        bind = &net.TCPAddr{
            IP:   net.IPv4(0, 0, 0, 0),
            Port: 0,
            Zone: "",
        }
    }

    dialer := net.Dialer{
        Deadline:  time.Now().Add(CONNECT_TIMEOUT),
        LocalAddr: bind,
        Control: func(network, address string, connection syscall.RawConn) (err error) {
            var operr error

            f := func(fd uintptr) {
                operr = setSocketOptions(fd)
            }

            if err := connection.Control(f); err != nil {
                return err
            } else {
                return operr
            }
        },
    }

    if connection, err := dialer.Dial("tcp4", address); err != nil {
        return nil, err
    } else if connection == nil {
        return nil, fmt.Errorf("invalid TCP socket (%v)", connection)
    } else {
        defer connection.Close()

        deadline := time.Now().Add(READ_TIMEOUT)
        buffer := make([]byte, 1024)

        if err := connection.SetDeadline(deadline); err != nil {
            return nil, err
        }

        if _, err := connection.Write(request); err != nil {
            return nil, fmt.Errorf("failed to write to UDP socket [%v]", err)
        }

        // NTS: set-ip doesn't return a reply
        if request[1] == 0x96 {
            return nil, nil
        }

        if N, err := connection.Read(buffer); err != nil {
            return nil, err
        } else {
            dump(buffer[:N])

            return buffer[0:N], nil
        }
    }
}

func listen(ch chan []byte) error {
    socket, err := net.ListenUDP("udp", listenAddr)
    if err != nil {
        return err
    }

    defer socket.Close()

    events := make(chan []byte)
    e := make(chan error)

    go func() {
        buffer := make([]byte, 1024)

        for {
            if N, _, err := socket.ReadFromUDP(buffer); err != nil {
                e <- err
            } else if N == 64 {
                m := make([]byte, 64)
                copy(m, buffer[0:N])
                events <- m
            }
        }
    }()

    for {
        select {
        case m := <-events:
            dump(m)
            ch <- m

        case err := <-e:
            return err
        }
    }

    return nil
}

func readAll(socket *net.UDPConn) ([][]byte, error) {
    reply := make(chan []byte)
    e := make(chan error)

    go func() {
        buffer := make([]byte, 1024)

        for {
            if N, _, err := socket.ReadFromUDP(buffer); err != nil {
                e <- err
            } else if N == 64 {
                m := make([]byte, 64)
                copy(m, buffer[0:N])
                reply <- m
            }
        }
    }()

    replies := [][]byte{}
    wait := time.After(2500 * time.Millisecond)

    for {
        select {
        case m := <-reply:
            dump(m)
            replies = append(replies, m)

        case <-wait:
            return replies, nil

        case err := <-e:
            return nil, err
        }
    }
}

func resolveAddr(address string) *net.UDPAddr {
    addr := netip.MustParseAddrPort(address)

    return net.UDPAddrFromAddrPort(addr)
}

func dump(packet []byte) {
    if debug {
        hex := "%02x %02x %02x %02x %02x %02x %02x %02x"

        for i := 0; i < 4; i++ {
            offset := i * 16
            u := packet[offset : offset+8]
            v := packet[offset+8 : offset+16]

            p := fmt.Sprintf(hex, u[0], u[1], u[2], u[3], u[4], u[5], u[6], u[7])
            q := fmt.Sprintf(hex, v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7])

            fmt.Printf("   %08x  %v  %v\n", offset, p, q)
        }

        fmt.Println()
    }
}
